modules/videogrid-workflowoperation/src/main/java/org/opencastproject/workflow/handler/videogrid/VideoGridWorkflowOperationHandler.java

```diff
@@ -0,0 +1,1007 @@
+   *          Time in milliseconds, e.g. 12567
+   * @return
+   *          Time in seconds, e.g. "12.567"
+   */
+  private String msToS(long timestamp)
+  {
+    double s = (double)timestamp / 1000;
+    return String.format(Locale.US, "%.3f", s);   // Locale.US to get a . instead of a ,
+  }
+
+  /**
+   * Finds and returns the first track matching the given id in a list of tracks
+   * @param trackId
+   *          The id of the track we're looking for
+   * @param tracks
+   *          The collection of tracks we're looking in
+   * @return
+   *          The first track with the given trackId
+   */
+  private Track getTrackByID(String trackId, List<Track> tracks) {
+    for (Track t : tracks) {
+      if (t.getIdentifier().contains(trackId)) {
+        logger.debug("Track-Id from smil found in Mediapackage ID: " + t.getIdentifier());
+        return t;
+      }
+    }
+    throw new IllegalStateException("No track matching smil Track-id: " + trackId);
+  }
+
+  /**
+   * Determine the largest dimension of the given list of tracks
+   *
+   * @param tracks
+   *          the list of tracks
+   * @param forceDivisible
+   *          Whether to enforce the track's dimension to be divisible by two
+   * @return the largest dimension from the list of track
+   */
+  private LayoutArea determineDimension(List<Track> tracks, boolean forceDivisible) {
+    Tuple<Track, LayoutArea> trackDimension = getLargestTrack(tracks);
+    if (trackDimension == null)
+      return null;
+
+    if (forceDivisible && (trackDimension.getB().getHeight() % 2 != 0 || trackDimension.getB().getWidth() % 2 != 0)) {
+      LayoutArea scaledDimension = new LayoutArea((trackDimension.getB().getWidth() / 2) * 2, (trackDimension
+              .getB().getHeight() / 2) * 2);
+      logger.info("Determined output dimension {} scaled down from {} for track {}", scaledDimension,
+              trackDimension.getB(), trackDimension.getA());
+      return scaledDimension;
+    } else {
+      logger.info("Determined output dimension {} for track {}", trackDimension.getB(), trackDimension.getA());
+      return trackDimension.getB();
+    }
+  }
+
+  /**
+   * Returns the track with the largest resolution from the list of tracks
+   *
+   * @param tracks
+   *          the list of tracks
+   * @return a {@link Tuple} with the largest track and it's dimension
+   */
+  private Tuple<Track, LayoutArea> getLargestTrack(List<Track> tracks) {
+    Track track = null;
+    LayoutArea dimension = null;
+    for (Track t : tracks) {
+      if (!t.hasVideo())
+        continue;
+
+      VideoStream[] videoStreams = TrackSupport.byType(t.getStreams(), VideoStream.class);
+      int frameWidth = videoStreams[0].getFrameWidth();
+      int frameHeight = videoStreams[0].getFrameHeight();
+      if (dimension == null || (frameWidth * frameHeight) > (dimension.getWidth() * dimension.getHeight())) {
+        dimension = new LayoutArea(frameWidth, frameHeight);
+        track = t;
```

Comment:
This might fail if the tracks have very strange dimensions, like very narrow, etc

Reply:
Maybe I'm blind, but you could explain to me why strange dimensions would make this fail?

Reply:
Maybe I did misunderstood some code, where this is information is used, since I don't find it any more. But what happens if e.g. there is one track ist is 1920x1080(16:9) and a track which is 2880x720(21:9). Both have the same area, but different dimensions. Just make sure it does not cause any issues when generating the grid.

Reply:
Thanks for explaining, I get it now. Should not be a problem though, as this function is used purely for gathering information if I recall correctly.

I just checked and apparently the only use of the method is when trying to determine the dimensions of a single track, so the size of the area does not actually factor in. The function is too powerful for what it is actually used for :')

Reply:
I understand @pascalseeland concerns, we found that some people have 21:9 monitors and if they record their screens you will have a very wide video, plus others.  H264 in their latest level could achieve more than 4096x2304 pixels. making a grid of videos. (This error can occur for example when you try to scale Tears of steel in their original aspect ratio to 4K).

It will be nice to handle such edge cases, FFmpeg will fail and tell you about this issue.